feat: Support standard environment variables

PSanetra opened the pull request:
Closes #81

## Changes

- Make previously required parameters of `oauth_channel.create_oauth2_client_credentials_channel()` and `oauth_channel.create_camunda_cloud_channel()` optional

## API Updates

### New Features *(required)*

`oauth_channel.create_oauth2_client_credentials_channel()` and `oauth_channel.create_camunda_cloud_channel()` support the following environment variables:
* `ZEEBE_ADDRESS`
* `CAMUNDA_CLUSTER_ID`
* `CAMUNDA_CLUSTER_REGION`
* `CAMUNDA_CLIENT_ID`
* `ZEEBE_CLIENT_ID`
* `CAMUNDA_CLIENT_SECRET`
* `ZEEBE_CLIENT_SECRET`
* `CAMUNDA_OAUTH_URL`
* `ZEEBE_AUTHORIZATION_SERVER_URL`
* `CAMUNDA_CREDENTIALS_SCOPES`
* `CAMUNDA_TOKEN_AUDIENCE`
* `ZEEBE_TOKEN_AUDIENCE`

## Checklist

- [x] Unit tests
- [ ] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda-community-hub/pyzeebe?pullRequest=509) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/camunda-community-hub/pyzeebe?pullRequest=509) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda-community-hub/pyzeebe?pullRequest=509) it.</sub>

felicijus commented:
@dimastbk I would like to check the Camunda Cloud implementation side of this again, review should be in tomorrow.

@PSanetra thanks you, you were quicker I also had this planned.

felicijus commented:
Let's wait for #514 and #510 to be closed.

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -4,9 +4,152 @@
+    return f"{cluster_id}.{cluster_region}.zeebe.camunda.io"
+
+
+def get_camunda_credentials_scopes(scopes: Optional[str]) -> Optional[str]:
+    """
+    Args:
+        scopes (str, optional): The camunda credentials scopes as single string.
+
+    Returns:
+        str: The camunda cedential scopes string.
+            Default: Value from CAMUNDA_CREDENTIALS_SCOPES
+    """
+
+    return scopes or os.getenv("CAMUNDA_CREDENTIALS_SCOPES")
```
`CAMUNDA_CREDENTIALS_SCOPES='Zeebe' is the scope of the Camunda Console credentiels, **not** the scope for authorization.

This is not to be used.

PSanetra replied:
Do you know about some documentation regarding this?

PSanetra replied:
According to the old code at https://github.com/camunda-community-hub/pyzeebe/pull/509/files#diff-24bc55faae9e3f2a8af987d3894672f979e646328ccd96c67b71f308a02f3e37L83 `Zeebe` was previously also used as the scope. So I am not sure what you mean.

felicijus replied:
I can't find it but the Camunda Docs only mention audience not scope and applying the scope to `CAMUNDA_CREDENTIALS_SCOPES` aka `Zeebe` fails with my own integration tests. #514 #511 #510

PSanetra replied:
I have removed the default value for the scopes parameters

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -4,9 +4,152 @@
+    return ret
+
+
+def get_camunda_cloud_cluster_id(cluster_id: Optional[str]) -> Optional[str]:
```
Keep it short `get_camunda_cluster_id`, because `CAMUNDA_CLUSTER_ID`.

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -4,9 +4,152 @@
+    return cluster_id or os.getenv("CAMUNDA_CLUSTER_ID")
+
+
+def get_camunda_cloud_cluster_region(cluster_region: Optional[str]) -> str:
```
Keep it short `get_camunda_cluster_region`, because `CAMUNDA_CLUSTER_REGION`.

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -4,9 +4,152 @@
+                     or camunda cloud token audience if camunda cluster_id is available
+    """
+
+    return (
+        token_audience
+        or os.getenv("CAMUNDA_TOKEN_AUDIENCE", os.getenv("ZEEBE_TOKEN_AUDIENCE"))
+        or get_camunda_cloud_hostname(None, None)
+    )
```
Does it have to be nested ?

````python
return (
        token_audience
        or os.getenv("CAMUNDA_TOKEN_AUDIENCE")
        or os.getenv("ZEEBE_TOKEN_AUDIENCE")
        or get_camunda_cloud_hostname(None, None)
    )
````

felicijus commented on pyzeebe/channel/oauth_channel.py:
```diff
@@ -116,34 +152,38 @@ def create_camunda_cloud_channel(
+    # NOTE: Overwrite the _oauth.fetch_token method to include client_id in the request body
+    oauth2_client_credentials._func_retrieve_token = partial(
+        oauth2_client_credentials._func_retrieve_token,
         include_client_id=True,
-        token_url=authorization_server,
-        client_secret=client_secret,
-        audience=audience,
     )
```
This is not working. Please don't change the partial. Revert.

PSanetra replied:
Do you know why? I am pretty new to python and as far as I have seen, those parameters were already applied to `oauth2_client_credentials._func_retrieve_token`.

felicijus replied:
Yes because it expects a partial in the definition and you are specifiying a partial with the partial expecting function. (circular)

Instead the oauth2session.fetch_token should be parametrized with the partial to prepare it to be executed before every (not really) rpc.

For more specifics look at the class `Oauth2ClientCredentialsMetadataPlugin`

felicijus replied:
@PSanetra You are correct with the testing, I could add some more specific unit tests for the partial, that is true. Though that won't solve the main problem.

The problem is the behaviour for specific parameters can really only be testet with integration tests, which for Camunda Cloud (SaaS) i do locally, but it isn't possible for this repo, because there is no pyzeebe Camunda Cloud Cluster...

PSanetra replied:
I have reverted this change

felicijus commented on pyzeebe/channel/oauth_channel.py:
```diff
@@ -53,12 +71,17 @@ def create_oauth2_client_credentials_channel(
-        audience=audience,
+        client_id=get_camunda_client_id(client_id),
+        client_secret=get_camunda_client_secret(client_secret),
+        authorization_server=authorization_server or "",
```
I don't get why `or ""`. 😅

PSanetra replied:
Just to make the type checker happy :sweat_smile: 
Otherwise the type would be `str | None`, but the parameter type is just `str`

felicijus replied:
Ah okay, but since it's not optional it should be set which means woever sets this parameter should not assign `None` (This would mean the `get_camunda_oauth_url`)

PSanetra replied:
@felicijus right, if this parameter is missing for some reason, a `ValueError` is already raised in line 77. Sadly the type checker does not consider this

felicijus replied:
Should the error not be rased in the `get_camunda_oauth_url` function?
Return type should be str not optional[str]

Also could you change to [PEP604](https://peps.python.org/pep-0604/) recent change from @dimastbk.

PSanetra replied:
The problem is that in case of `create_camunda_cloud_channel()` it needs to be possible to use `"https://login.cloud.camunda.io/oauth/token"` as additional fallback. So `get_camunda_oauth_url` should not throw there.

PSanetra replied:
Is the [PEP604](https://peps.python.org/pep-0604/) change already in master?

felicijus replied:
Yes. You may need to sync your branch.

PSanetra replied:
I have rebased the branch and applied PEP604

felicijus replied:
Can you revert the changes to the partial too ?

PSanetra replied:
Done, see https://github.com/camunda-community-hub/pyzeebe/pull/509/commits/12535fc372f7bd387be09deff82ae8977d6da396

felicijus reviewed (changes requested):
Looks promising. 🚀

felicijus commented:
I will review again 🚀

PSanetra commented:
@felicijus @dimastbk do you have time for another review?

felicijus commented:
Sorry forgot to finish mine...

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,140 @@
+    return cluster_region or os.getenv("CAMUNDA_CLUSTER_REGION") or "bru-2"
+
+
+def get_camunda_token_audience(token_audience: str | None) -> str | None:
+    """
+    Args:
+        token_audience (str, optional): The token audience provided as parameter.
+
+    Returns:
+        str: The token audience.
+            Default: Value from CAMUNDA_TOKEN_AUDIENCE
+                     or ZEEBE_TOKEN_AUDIENCE environment variable
+                     or camunda cloud token audience if camunda cluster_id is available
+    """
+
+    return (
+        token_audience
+        or os.getenv("CAMUNDA_TOKEN_AUDIENCE")
+        or os.getenv("ZEEBE_TOKEN_AUDIENCE")
+        or get_camunda_cloud_hostname(None, None)
+    )
```
Currently the only ways I know is either a different sentinal object than `None`, we could have `Empty`.
@PSanetra please think of a better name.

Or we could use a wrapper to solve this.

I have to check about typing but a custom sentinel seems the better approach.

felicijus commented on pyzeebe/channel/oauth_channel.py:
```diff
@@ -5,18 +5,28 @@
 
 from pyzeebe.channel.channel_options import get_channel_options
+from pyzeebe.channel.utils import (
+    get_camunda_client_id,
+    get_camunda_client_secret,
+    get_camunda_cloud_hostname,
+    get_camunda_cluster_id,
+    get_camunda_cluster_region,
+    get_camunda_oauth_url,
+    get_camunda_token_audience,
+    get_zeebe_address,
+)
 from pyzeebe.credentials.oauth import Oauth2ClientCredentialsMetadataPlugin
 from pyzeebe.types import ChannelArgumentType
 
 
 def create_oauth2_client_credentials_channel(
-    grpc_address: str,
-    client_id: str,
-    client_secret: str,
-    authorization_server: str,
+    grpc_address: str | None = None,
+    client_id: str | None = None,
+    client_secret: str | None = None,
```
Using the sentinal is a bigger change than expected...
I think defining the defaults outside of the helper functions, makes the code more concise. And defines them where they are applied.

I will just leave my whole file here.

````python
from __future__ import annotations

from functools import partial

import grpc

from pyzeebe.channel.channel_options import get_channel_options
from pyzeebe.channel.utils import (
    get_camunda_address,
    get_camunda_client_id,
    get_camunda_client_secret,
    get_camunda_cluster_id,
    get_camunda_cluster_region,
    get_camunda_oauth_url,
    get_camunda_token_audience,
    get_zeebe_address,
)
from pyzeebe.credentials.oauth import Oauth2ClientCredentialsMetadataPlugin
from pyzeebe.types import ChannelArgumentType, Unset


def create_oauth2_client_credentials_channel(
    grpc_address: str = Unset,
    client_id: str = Unset,
    client_secret: str = Unset,
    authorization_server: str = Unset,
    scope: str | None = Unset,
    audience: str | None = Unset,
    channel_credentials: grpc.ChannelCredentials = grpc.ssl_channel_credentials(),
    channel_options: ChannelArgumentType | None = None,
    leeway: int = 60,
    expire_in: int | None = None,
) -> grpc.aio.Channel:
    """Create a gRPC channel for connecting to Camunda 8 (Self-Managed) with OAuth2ClientCredentials.

    https://oauth.net/2/grant-types/client-credentials/
    https://datatracker.ietf.org/doc/html/rfc6749#section-11.2.2

    Args:
        grpc_address (str | None): Zeebe Gateway Address.
            Defaults to value from ZEEBE_ADDRESS environment variable
                    or "{CAMUNDA_CLUSTER_ID}.{CAMUNDA_CLUSTER_REGION}.zeebe.camunda.io:443"
                    or "localhost:26500".
        client_id (str | None): The client id.
            Defaults to value from CAMUNDA_CLIENT_ID or ZEEBE_CLIENT_ID environment variable
        client_secret (str | None): The client secret.
            Defaults to value from CAMUNDA_CLIENT_SECRET or ZEEBE_CLIENT_SECRET environment variable
        authorization_server (str | None): The authorization server issuing access tokens
            to the client after successfully authenticating the client.
            Defaults to value from CAMUNDA_OAUTH_URL or ZEEBE_AUTHORIZATION_SERVER_URL environment variable
        scope (str | None): The scope of the access request.
        audience (str | None): The audience for authentication.
            Defaults to value from CAMUNDA_TOKEN_AUDIENCE or ZEEBE_TOKEN_AUDIENCE environment variable

        channel_credentials (grpc.ChannelCredentials): The gRPC channel credentials.
            Defaults to grpc.ssl_channel_credentials().
        channel_options (ChannelArgumentType | None): Additional options for the gRPC channel.
            Defaults to None.
            See https://grpc.github.io/grpc/python/glossary.html#term-channel_arguments

        leeway (int): The number of seconds to consider the token as expired before the actual expiration time.
            Defaults to 60.
        expire_in (int | None): The number of seconds the token is valid for. Defaults to None.
            Should only be used if the token does not contain an "expires_in" attribute.

    Returns:
        grpc.aio.Channel: A gRPC channel connected to the Zeebe Gateway.

    Raises:
        InvalidOAuthCredentialsError: One of the provided camunda credentials is not correct
    """

    if grpc_address is Unset:
        grpc_address = get_zeebe_address()

    if client_id is Unset:
        client_id = get_camunda_client_id()

    if client_secret is Unset:
        client_secret = get_camunda_client_secret()

    if authorization_server is Unset:
        authorization_server = get_camunda_oauth_url()

    if scope is Unset:
        scope = None

    if audience is Unset:
        audience = get_camunda_token_audience()

    oauth2_client_credentials = Oauth2ClientCredentialsMetadataPlugin(
        client_id=client_id,
        client_secret=client_secret,
        authorization_server=authorization_server,
        scope=scope,
        audience=audience,
        leeway=leeway,
        expire_in=expire_in,
    )

    call_credentials: grpc.CallCredentials = grpc.metadata_call_credentials(oauth2_client_credentials)
    composite_credentials: grpc.ChannelCredentials = grpc.composite_channel_credentials(
        channel_credentials or grpc.ssl_channel_credentials(), call_credentials
    )

    channel: grpc.aio.Channel = grpc.aio.secure_channel(
        target=grpc_address,
        credentials=composite_credentials,
        options=get_channel_options(channel_options),
    )

    return channel


def create_camunda_cloud_channel(
    client_id: str = Unset,
    client_secret: str = Unset,
    cluster_id: str = Unset,
    region: str = Unset,
    authorization_server: str = Unset,
    scope: str | None = Unset,
    audience: str | None = Unset,
    channel_credentials: grpc.ChannelCredentials = grpc.ssl_channel_credentials(),
    channel_options: ChannelArgumentType | None = None,
    leeway: int = 60,
    expire_in: int | None = None,
) -> grpc.aio.Channel:
    """Create a gRPC channel for connecting to Camunda 8 Cloud (SaaS).

    Args:
        client_id (str | None): The client id.
            Defaults to value from CAMUNDA_CLIENT_ID or ZEEBE_CLIENT_ID environment variable.
        client_secret (str | None): The client secret.
            Defaults to value from CAMUNDA_CLIENT_SECRET or ZEEBE_CLIENT_SECRET environment variable.
        cluster_id (str | None): The ID of the cluster to connect to.
            Defaults to value from CAMUNDA_CLUSTER_ID environment variable.
        region (str | None): The region of the cluster.
            Defaults to value from CAMUNDA_CLUSTER_REGION environment variable or 'bru-2'.
        authorization_server (str | None): The authorization server issuing access tokens
            to the client after successfully authenticating the client.
            Defaults to value from CAMUNDA_OAUTH_URL
                or ZEEBE_AUTHORIZATION_SERVER_URL environment variable
                or "https://login.cloud.camunda.io/oauth/token".
        scope (str | None): The scope of the access request.
        audience (str | None): The audience for authentication.
            Defaults to value from CAMUNDA_TOKEN_AUDIENCE
                or ZEEBE_TOKEN_AUDIENCE environment variable
                or "{cluster_id}.{region}.zeebe.camunda.io"
                or "zeebe.camunda.io".

        channel_credentials (grpc.ChannelCredentials): The gRPC channel credentials.
            Defaults to grpc.ssl_channel_credentials().
        channel_options (ChannelArgumentType | None): Additional options for the gRPC channel.
            Defaults to None.
            See https://grpc.github.io/grpc/python/glossary.html#term-channel_arguments

        leeway (int): The number of seconds to consider the token as expired before the actual expiration time.
            Defaults to 60.
        expire_in (int | None): The number of seconds the token is valid for. Defaults to None.
            Should only be used if the token does not contain an "expires_in" attribute.

    Returns:
        grpc.aio.Channel: The gRPC channel for connecting to Camunda Cloud.
    """

    if client_id is Unset:
        client_id = get_camunda_client_id()

    if client_secret is Unset:
        client_secret = get_camunda_client_secret()

    if cluster_id is Unset:
        cluster_id = get_camunda_cluster_id()

    if region is Unset:
        region = get_camunda_cluster_region() or "bru-2"

    if authorization_server is Unset:
        authorization_server = get_camunda_oauth_url() or "https://login.cloud.camunda.io/oauth/token"

    if scope is Unset:
        scope = None

    if audience is Unset:
        audience = get_camunda_token_audience() or "zeebe.camunda.io"

    oauth2_client_credentials = Oauth2ClientCredentialsMetadataPlugin(
        client_id=client_id,
        client_secret=client_secret,
        authorization_server=authorization_server,
        scope=scope,
        audience=audience,
        leeway=leeway,
        expire_in=expire_in,
    )

    # NOTE: Overwrite the _oauth.fetch_token method to include client_id, client_secret in the request body
    func = partial(
        oauth2_client_credentials._oauth.fetch_token,
        include_client_id=True,
        token_url=authorization_server,
        client_secret=client_secret,
        audience=audience,
    )
    oauth2_client_credentials._func_retrieve_token = func

    call_credentials: grpc.CallCredentials = grpc.metadata_call_credentials(oauth2_client_credentials)
    composite_credentials: grpc.ChannelCredentials = grpc.composite_channel_credentials(
        channel_credentials or grpc.ssl_channel_credentials(), call_credentials
    )

    grpc_address = get_camunda_address(cluster_id=cluster_id, cluster_region=region)

    channel: grpc.aio.Channel = grpc.aio.secure_channel(
        target=grpc_address, credentials=composite_credentials, options=get_channel_options(channel_options)
    )

    return channel
````

felicijus reviewed (changes requested):
This should cover everything for now.

If we use a sentinal like this:
````python
Unset = str("UNSET")
````

We solve the problem of `None` as a valid value and pseudo Optional arguments, meaning we can eliminate them.
This means for real Optional arguments the User can define `None` as a valid value. 🤖

PSanetra commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,140 @@
+    return cluster_id or os.getenv("CAMUNDA_CLUSTER_ID")
+
+
+def get_camunda_cluster_region(cluster_region: str | None) -> str:
+    """
+    Args:
+        cluster_region (str, optional): The camunda cluster region provided as parameter.
+
+    Returns:
+        str: The camunda cluster region.
+            Default: Value from CAMUNDA_CLUSTER_REGION environment variable or 'bru-2'
+    """
+
+    return cluster_region or os.getenv("CAMUNDA_CLUSTER_REGION") or "bru-2"
```
I think we can not raise an error here as the function is used like this: `region = get_camunda_cluster_region() or "bru-2"`

PSanetra commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,140 @@
+    return cluster_region or os.getenv("CAMUNDA_CLUSTER_REGION") or "bru-2"
+
+
+def get_camunda_token_audience(token_audience: str | None) -> str | None:
+    """
+    Args:
+        token_audience (str, optional): The token audience provided as parameter.
+
+    Returns:
+        str: The token audience.
+            Default: Value from CAMUNDA_TOKEN_AUDIENCE
+                     or ZEEBE_TOKEN_AUDIENCE environment variable
+                     or camunda cloud token audience if camunda cluster_id is available
+    """
+
+    return (
+        token_audience
+        or os.getenv("CAMUNDA_TOKEN_AUDIENCE")
+        or os.getenv("ZEEBE_TOKEN_AUDIENCE")
+        or get_camunda_cloud_hostname(None, None)
+    )
```
I think we can not raise an error here as the function is used like this: `audience = get_camunda_token_audience() or "zeebe.camunda.io"`

PSanetra commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,140 @@
+    return os.getenv("ZEEBE_ADDRESS", camunda_cloud_address or DEFAULT_ZEEBE_ADDRESS)
+
+
+def get_camunda_oauth_url(oauth_url: str | None) -> str | None:
+    """
+    Args:
+        oauth_url (str, optional): The camunda platform authorization server url provided as parameter.
+
+    Returns:
+        str: The camunda platform authorization server url.
+            Default: Value from CAMUNDA_OAUTH_URL or ZEEBE_AUTHORIZATION_SERVER_URL environment variable
+    """
+    return oauth_url or os.getenv("CAMUNDA_OAUTH_URL", os.getenv("ZEEBE_AUTHORIZATION_SERVER_URL"))
```
I think we can not raise an error here as the function is used like this: `authorization_server = get_camunda_oauth_url() or "https://login.cloud.camunda.io/oauth/token"`

PSanetra commented:
@felicijus I have applied your latest suggestions, but needed to make some adjustments. E.g. the functions `get_zeebe_address`, `get_camunda_oauth_url`, `get_camunda_cluster_region` and `get_camunda_token_audience` need to have an optional `default_value` argument, which need to be considered as fallback in the case of usages in `create_camunda_cloud_channel`,

Also I have seen that version `4.0.0` was released, so I think the change of argument order in `create_camunda_cloud_channel` and probably also the introduction of the new `Unset` default values need to be considered as breaking changes.

felicijus commented:
Thank you. I will look into it this afternoon.

felicijus commented:
Docs will be done by you @PSanetra ?

PSanetra commented:
@felicijus I can do that, but I am not sure where to put them. Do you have a suggestion? I think it should generally be just documentation of the supported environment variables, right?

felicijus commented:
Yes general support with what is supported as a seperate section, and add a short description to each channel creating function and the defaults used.

About the breaking changes, yes and no... At least for the oauth channel creation - no. There was no support for Environment variables at all and you also had no imposter `None` as default. ✅ 
For `create_secure_channel` and `create_insecure_channel` it can be considered a breaking change. ⚠️ 

So I guess 4.1.0 or whatever @dimastbk decides.

felicijus reviewed (approved):
✅

PSanetra commented:
@felicijus I have added some documentation can you review it? I did not write reStructuredText docs before.

I have also added some tests which should improve coverage.

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
-    if grpc_address:
-        return grpc_address
-    return os.getenv("ZEEBE_ADDRESS", DEFAULT_ZEEBE_ADDRESS)
+def get_zeebe_address(default_value: str | None = None) -> str:
+    """
+    Args:
+        default_value (str, optional): Default value to be used if no other value was discovered.
+    Returns: Value from ZEEBE_ADDRESS environment variable
+            or provided default_value
+            or "localhost:26500"
+    """
+
+    return os.getenv("ZEEBE_ADDRESS") or default_value or DEFAULT_ZEEBE_ADDRESS
```
In the documentation only ever default is mentioned not default_value, also l like default more (shorter, simple).

Also docstring changes to make it compatible with sphinx. Like `Returns` with a type or in the other functions `Raises`.

```suggestion
def get_zeebe_address(default: str = None) -> str:
    """
    Get the Zeebe Gateway Address.

    Args:
        default (str, optional): Default value to be used if no other value was discovered.

    Returns:
        str: ZEEBE_ADDRESS environment variable or provided default or "localhost:26500"
    """
    return os.getenv("ZEEBE_ADDRESS") or default or DEFAULT_ZEEBE_ADDRESS
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+    return os.getenv("ZEEBE_ADDRESS") or default_value or DEFAULT_ZEEBE_ADDRESS
+
+
+def get_camunda_oauth_url(default_value: str | None = None) -> str:
+    """
+    Args:
+        default_value (str, optional): Default value to be used if no other value was discovered.
+    Returns: The camunda platform authorization server url.
+            Default: Value from CAMUNDA_OAUTH_URL
+                    or ZEEBE_AUTHORIZATION_SERVER_URL environment variable
+                    or provided default_value
+    """
+    r = os.getenv("CAMUNDA_OAUTH_URL") or os.getenv("ZEEBE_AUTHORIZATION_SERVER_URL") or default_value
```
```suggestion
def get_camunda_oauth_url(default: str = None) -> str:
    """
    Get the Camunda OAuth URL or Zeebe Authorization Server URL.

    Args:
        default (str, optional): Default value to be used if no other value was discovered.

    Returns:
        str: CAMUNDA_OAUTH_URL or ZEEBE_AUTHORIZATION_SERVER_URL environment variable or provided default

    Raises:
        EnvironmentError: If neither CAMUNDA_OAUTH_URL nor ZEEBE_AUTHORIZATION_SERVER_URL is provided.
    """
    r = os.getenv("CAMUNDA_OAUTH_URL") or os.getenv("ZEEBE_AUTHORIZATION_SERVER_URL") or default
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+
+
+def get_camunda_client_id() -> str:
+    """
+    Returns: The client id.
+            Default: Value from CAMUNDA_CLIENT_ID or ZEEBE_CLIENT_ID environment variable
+    """
```
```suggestion
    """
    Get the Camunda Client ID.

    Returns:
        str: CAMUNDA_CLIENT_ID or ZEEBE_CLIENT_ID environment variable

    Raises:
        EnvironmentError: If neither CAMUNDA_CLIENT_ID nor ZEEBE_CLIENT_ID is provided.
    """
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+
+
+def get_camunda_client_secret() -> str:
+    """
+    Returns:
+        str: The client secret.
+            Default: Value from CAMUNDA_CLIENT_SECRET or ZEEBE_CLIENT_SECRET environment variable
+    """
```
```suggestion
    """
    Get the Camunda Client Secret.

    Returns:
        str: CAMUNDA_CLIENT_SECRET or ZEEBE_CLIENT_SECRET environment variable

    Raises:
        EnvironmentError: If neither CAMUNDA_CLIENT_SECRET nor ZEEBE_CLIENT_SECRET is provided.
    """
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+
+
+def get_camunda_cluster_id() -> str:
+    """
+    Returns: The camunda cluster id.
+            Default: Value from CAMUNDA_CLUSTER_ID environment variable
+    """
```
```suggestion
    """
    Get the Camunda Cluster ID.

    Returns:
        str: CAMUNDA_CLUSTER_ID environment variable

    Raises:
        EnvironmentError: If CAMUNDA_CLUSTER_ID is not provided.
    """
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+    return r
+
+
+def get_camunda_cluster_region(default_value: str | None = None) -> str:
+    """
+    Args:
+        default_value (str, optional): Default value to be used if no other value was discovered.
+    Returns: The camunda cluster region.
+            Default: Value from CAMUNDA_CLUSTER_REGION environment variable
+    """
+
+    r = os.getenv("CAMUNDA_CLUSTER_REGION") or default_value
```
```suggestion
def get_camunda_cluster_region(default: str = None) -> str:
    """
    Get the Camunda Cluster Region.

    Args:
        default (str, optional): Default value to be used if no other value was discovered.

    Returns:
        str: CAMUNDA_CLUSTER_REGION environment variable or provided default

    Raises:
        EnvironmentError: If CAMUNDA_CLUSTER_REGION is not provided.
    """
    r = os.getenv("CAMUNDA_CLUSTER_REGION") or default
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+    return r
+
+
+def get_camunda_token_audience(default_value: str | None = None) -> str:
+    """
+    Args:
+        default_value (str, optional): Default value to be used if no other value was discovered.
+    Returns: The token audience.
+            Default: Value from CAMUNDA_TOKEN_AUDIENCE
+                     or ZEEBE_TOKEN_AUDIENCE environment variable
+                     or provided default_value
+    """
+    r = os.getenv("CAMUNDA_TOKEN_AUDIENCE") or os.getenv("ZEEBE_TOKEN_AUDIENCE") or default_value
```
```suggestion
def get_camunda_token_audience(default: str = None) -> str:
    """
    Get the Camunda Token Audience.

    Args:
        default (str, optional): Default value to be used if no other value was discovered.

    Returns:
        str: CAMUNDA_TOKEN_AUDIENCE or ZEEBE_TOKEN_AUDIENCE environment variable or provided default

    Raises:
        EnvironmentError: If neither CAMUNDA_TOKEN_AUDIENCE nor ZEEBE_TOKEN_AUDIENCE is provided.
    """
    r = os.getenv("CAMUNDA_TOKEN_AUDIENCE") or os.getenv("ZEEBE_TOKEN_AUDIENCE") or default
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+
+
+def get_camunda_address(cluster_id: str | None = None, cluster_region: str | None = None) -> str:
+    """
+    Args:
+        cluster_id (str, optional): The camunda cluster id provided as parameter.
+            Default: Value from CAMUNDA_CLUSTER_ID environment variable or None
+        cluster_region (str, optional): The camunda cluster region provided as parameter.
+            Default: Value from CAMUNDA_CLUSTER_REGION environment variable.
+    Returns:
+        str: The Camunda Cloud grpc server address.
+    """
```
```suggestion
    """
    Get the Camunda Cloud gRPC server address.

    Args:
        cluster_id (str, optional): The Camunda cluster ID provided as parameter.
            Default: Value from CAMUNDA_CLUSTER_ID environment variable or None.
        cluster_region (str, optional): The Camunda cluster region provided as parameter.
            Default: Value from CAMUNDA_CLUSTER_REGION environment variable.

    Returns:
        str: The Camunda Cloud gRPC server address.

    Raises:
        EnvironmentError: If either cluster_id or cluster_region is not provided.
    """
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+    if (cluster_id is None) or (cluster_region is None):
+        raise EnvironmentError("The cluster_id and cluster_region must be provided!")
+
+    # if (cluster_id is not None) and (cluster_region is not None):  # and
```
```suggestion
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+    Returns:
+        str: The Camunda Cloud grpc server address.
+    """
+
```
```suggestion
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+    Returns: The camunda cluster id.
+            Default: Value from CAMUNDA_CLUSTER_ID environment variable
+    """
+
```
```suggestion
```

felicijus commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,123 @@
+        str: The client secret.
+            Default: Value from CAMUNDA_CLIENT_SECRET or ZEEBE_CLIENT_SECRET environment variable
+    """
+
```
```suggestion
```

felicijus reviewed (changes requested):
I suggested some changes...

Please add this to the end of `docs/channels_reference.rst` to complete Channel Reference.

````
Utilities (Environment) 
-----------------------

.. autofunction:: pyzeebe.channel.utils.get_zeebe_address

.. autofunction:: pyzeebe.channel.utils.get_camunda_oauth_url

.. autofunction:: pyzeebe.channel.utils.get_camunda_client_id

.. autofunction:: pyzeebe.channel.utils.get_camunda_client_secret

.. autofunction:: pyzeebe.channel.utils.get_camunda_cluster_id

.. autofunction:: pyzeebe.channel.utils.get_camunda_cluster_region

.. autofunction:: pyzeebe.channel.utils.get_camunda_token_audience

.. autofunction:: pyzeebe.channel.utils.get_camunda_address
````

PSanetra commented:
@felicijus Alright I have applied the suggestions

felicijus commented on docs/channels_configuration.rst:
```diff
@@ -0,0 +1,120 @@
+  :Description:  
+    Specifies the URL of the authorization server issuing access tokens to the client.
+  
+  :Usage:        
+    Used in the OAuth2 and Camunda Cloud channel configurations.
+  
+  :Default:      
+    `"https://login.cloud.camunda.io/oauth/token"` if not specified.
+    :func:`pyzeebe.create_camunda_cloud_channel`.
+
+**CAMUNDA_CLUSTER_ID**
+  :Description:  
+    The unique identifier for the Camunda Cloud cluster to connect to.
+  
+  :Usage:        
+    Required if channel initialization argument was not specified.
+    :func:`pyzeebe.create_camunda_cloud_channel`.
+
+**CAMUNDA_CLUSTER_REGION**
+  :Description:  
+    The region where the Camunda Cloud cluster is hosted.
+  
+  :Usage:        
+    Required for Camunda Cloud channels.
+  
+  :Default:      
+    `"bru-2"` if not provided.
+    :func:`pyzeebe.create_camunda_cloud_channel`.
+
+**CAMUNDA_TOKEN_AUDIENCE** / **ZEEBE_TOKEN_AUDIENCE**
+  :Description:  
+    Specifies the audience for the OAuth2 token.
+  
+  :Usage:        
+    Used when creating OAuth2 or Camunda Cloud channels.
+  
+  :Default:      
+    `"zeebe.camunda.io"` if not provided.
+    :func:`pyzeebe.create_camunda_cloud_channel`.
+
```
Missed something...

```suggestion
======================
Channels Configuration
======================

This document describes the environment variables and configurations used for establishing different gRPC channel connections to Camunda (Zeebe) instances, either with or without authentication.

Environment Variables
---------------------

The following environment variables are used to configure channels. The variables are grouped according to their relevance and usage context in each type of channel.

These variables are only considered if a corresponding argument was not passed (Unset) during initialization of a channel.

Common Variables
----------------

This variables is used across all types of channels:

**ZEEBE_ADDRESS**
  :Description:  
    The default address of the Zeebe Gateway.
  
  :Usage:        
    Used in both secure and insecure channel configurations.
    :func:`pyzeebe.create_insecure_channel`
    :func:`pyzeebe.create_secure_channel`
  
  :Default:      
    ``"localhost:26500"``

Common OAuth2 Variables
-----------------------

These variables are specifically for connecting to generic OAuth2 or Camunda Cloud instances.

**CAMUNDA_CLIENT_ID** / **ZEEBE_CLIENT_ID**
  :Description:  
    The client ID required for OAuth2 client credential authentication.
  
  :Usage:        
    Required for OAuth2 and Camunda Cloud channels.
    :func:`pyzeebe.create_oauth2_client_credentials_channel`
    :func:`pyzeebe.create_camunda_cloud_channel`

**CAMUNDA_CLIENT_SECRET** / **ZEEBE_CLIENT_SECRET**
  :Description:  
    The client secret for the OAuth2 client.
  
  :Usage:        
    Required for OAuth2 and Camunda Cloud channels.
    :func:`pyzeebe.create_oauth2_client_credentials_channel`
    :func:`pyzeebe.create_camunda_cloud_channel`

OAuth2 Variables (Self-Managed)
-------------------------------

These variables are primarily used for OAuth2 authentication in self-managed Camunda 8 instances.

**CAMUNDA_OAUTH_URL** / **ZEEBE_AUTHORIZATION_SERVER_URL**
  :Description:  
    Specifies the URL of the authorization server issuing access tokens to the client.
  
  :Usage:        
    Required if channel initialization argument was not specified.
    :func:`pyzeebe.create_oauth2_client_credentials_channel`

**CAMUNDA_TOKEN_AUDIENCE** / **ZEEBE_TOKEN_AUDIENCE**
  :Description:  
    Specifies the audience for the OAuth2 token.
  
  :Usage:        
    Used when creating OAuth2 or Camunda Cloud channels.
    :func:`pyzeebe.create_oauth2_client_credentials_channel`
  
  :Default:      
    ``None`` if not provided.

Camunda Cloud Variables (SaaS)
------------------------------

These variables are specifically for connecting to Camunda Cloud instances.

**CAMUNDA_OAUTH_URL** / **ZEEBE_AUTHORIZATION_SERVER_URL**
  :Description:  
    Specifies the URL of the authorization server issuing access tokens to the client.
  
  :Usage:        
    Used in the OAuth2 and Camunda Cloud channel configurations.
    :func:`pyzeebe.create_camunda_cloud_channel`
  
  :Default:      
    ``"https://login.cloud.camunda.io/oauth/token"`` if not specified.

**CAMUNDA_CLUSTER_ID**
  :Description:  
    The unique identifier for the Camunda Cloud cluster to connect to.
  
  :Usage:        
    Required if channel initialization argument was not specified.
    :func:`pyzeebe.create_camunda_cloud_channel`

**CAMUNDA_CLUSTER_REGION**
  :Description:  
    The region where the Camunda Cloud cluster is hosted.
  
  :Usage:        
    Required for Camunda Cloud channels.
    :func:`pyzeebe.create_camunda_cloud_channel`
  
  :Default:      
    ``"bru-2"`` if not provided.

**CAMUNDA_TOKEN_AUDIENCE** / **ZEEBE_TOKEN_AUDIENCE**
  :Description:  
    Specifies the audience for the OAuth2 token.
  
  :Usage:        
    Used when creating OAuth2 or Camunda Cloud channels.
    :func:`pyzeebe.create_camunda_cloud_channel`
  
  :Default:      
    ``"zeebe.camunda.io"`` if not provided.

```

felicijus reviewed (changes requested):
Missed some rst formatting.

Inline code:
````markup
``"localhost:26500"``
````

After that we are done I think. ✅

PSanetra commented:
@felicijus Alright I have applied that suggestion as well. I think when you want to merge this PR it makes sense to squash all the commit into one.

felicijus reviewed (approved):
Nice one. ✅

@dimastbk Please have a look and see if this can be released in 4.0.x...

dimastbk commented on pyzeebe/channel/utils.py:
```diff
@@ -5,9 +5,153 @@
+    Get the Camunda Cloud gRPC server address.
+
+    Args:
+        cluster_id (str, optional): The Camunda cluster ID provided as parameter.
+            Default: Value from CAMUNDA_CLUSTER_ID environment variable or None.
+        cluster_region (str, optional): The Camunda cluster region provided as parameter.
+            Default: Value from CAMUNDA_CLUSTER_REGION environment variable.
```
```suggestion
        cluster_id (str, optional): The Camunda cluster ID provided as parameter.
        cluster_region (str, optional): The Camunda cluster region provided as parameter.
```

PSanetra commented on pyzeebe/channel/oauth_channel.py:
```diff
@@ -5,52 +5,89 @@
-        authorization_server (str): The authorization server issuing access tokens
+        grpc_address (str, optional): Zeebe Gateway Address.
+            Defaults to value from ZEEBE_ADDRESS environment variable
+                    or "{CAMUNDA_CLUSTER_ID}.{CAMUNDA_CLUSTER_REGION}.zeebe.camunda.io:443"
+                    or "localhost:26500".
```
I am not sure this error is covered by any linter in the CI at the moment. Where do you see this?

dimastbk replied:
https://github.com/camunda-community-hub/pyzeebe/actions/runs/11719689122/job/32644814012?pr=509#step:7:40

or

```
cd docs
make html
```

PSanetra commented:
@dimastbk I have rebased the PR onto master. I have also fixed the linting issues, found in the CI, but I am not sure were you found the errors, regarding the indentation in the docs.

PSanetra commented:
@dimastbk I have resolved the remaining indentation issues.
